lib/curl_sha512_256.c

```diff
@@ -0,0 +1,588 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Evgeny Grin (Karlson2k), <k2k@narod.ru>.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if !defined(CURL_DISABLE_DIGEST_AUTH) && !defined(CURL_DISABLE_SHA512_256)
+
+#include "curl_sha512_256.h"
+#include "warnless.h"
+
+/* ** This implementation of SHA-512/256 hash calculation was originally ** *
+ * ** written by Evgeny Grin (Karlson2k) for GNU libmicrohttpd.          ** *
+ * ** The author ported the code to libcurl. The ported code is provided ** *
+ * ** under curl license.                                                ** *
+ * ** This is a minimal version with minimal optimisations. Performance  ** *
+ * ** can be significantly improved. Big-endian store and load macros    ** *
+ * ** are obvious targets for optimisation.                              ** */
+
+#if !defined(MHDX_UINT64) && defined(SIZEOF_LONG)
+#if SIZEOF_LONG == 8
+#define MHDX_UINT64 unsigned long
+#define MHDX_UINT64_C(val) val##UL
+#endif /* SIZEOF_LONG == 8 */
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64) && defined(HAVE_LONGLONG)
+#define MHDX_UINT64 unsigned long long
+#define MHDX_UINT64_C(val) val##ULL
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64)
+#define MHDX_UINT64 unsigned __int64
+#define MHDX_UINT64_C(val) ((MHDX_UINT64)val)
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#ifndef MHDX_UINT64_TRIM
+/* Trim possibly larger value to 64 bits.
+   Assuming that no type is larger than 64 bits. */
+#define MHDX_UINT64_TRIM(x) ((MHDX_UINT64)(x))
+#endif /* !MHDX_UINT64_TRIM */
```

Comment:
I propose you instead use the fact that we already have a 64 bit type figured out:

```suggestion
typedef unsigned CURL_TYPEOF_CURL_OFF_T mdhx_t
```

We could probably even consider making this type available for the entire libcurl, maybe as `curl_offu_t` or something?

Reply:
If `CURL_TYPEOF_CURL_OFF_T` is always a macro and never typedef it will work.
I'd suggest to reproduce standard `stdint.h`, 
``` C
typedef unsigned CURL_TYPEOF_CURL_OFF_T curl_uint64_t
#define CURL_UINT64_SUFFIX CURL_SUFFIX_CURL_OFF_TU
#define CURL_UINT64_C(val)  val##CURL_UINT64_SUFFIX
```
It would be quite easy to read and understand by newcomers.

Reply:
Note: for external API libcurl *may* use `CURL_TYPEOF_CURL_OFF_T` as 32 bit.
See https://github.com/curl/curl/blob/476adfeac019edec71d67e144f184f4dbf1a46fb/include/curl/system.h#L364 and https://github.com/curl/curl/blob/476adfeac019edec71d67e144f184f4dbf1a46fb/include/curl/system.h#L406
A assume that internally libcurl uses the same types.
So probably it would make sense to add something like 
``` C
#ifdef CURL_HAVE_CULT_OFF_T_64BIT
  typedef unsigned CURL_TYPEOF_CURL_OFF_T curl_uint64_t
#  define CURL_UINT64_SUFFIX CURL_SUFFIX_CURL_OFF_TU
#  define CURL_UINT64_C(val)  val##CURL_UINT64_SUFFIX
#endif

Reply:
Of maybe even
``` C
#ifdef CURL_HAVE_CULT_OFF_T_64BIT
  typedef unsigned CURL_TYPEOF_CURL_OFF_T curl_uint64_t
#  define CURL_HAVE_UINT64
#  define CURL_UINT64_SUFFIX CURL_SUFFIX_CURL_OFF_TU
#  define CURL_UINT64_C(val)  val##CURL_UINT64_SUFFIX
#endif
````
The code requiring 64 bits can be disabled by testing macro CURL_HAVE_UINT64

Reply:
> Note: for external API libcurl may use CURL_TYPEOF_CURL_OFF_T as 32 bit.

No, it only appears so (because the `system.h` header has not been updated). We have checks in the build that makes it fail if `curl_off_t` is not 64 bit. We can write code and rely on it being at least 64 bit. We require a working 64 bit datatype for building curl since a while back.

`curl_uint64_t` sounds excellent to my ears!

Reply:
> We can write code and rely on it being at least 64 bit. We require a working 64 bit datatype for building curl since a while back.
> 
> `curl_uint64_t` sounds excellent to my ears!

OK. Where to put this typedef? Which header is the best?
I think it should be internal-only and not to be exposed for public API.

Reply:
Right, let's not make it public until there is a need for that. I think you can put it in `lib/curl_setup.h`, which is a pretty generic and widely included header file for the library build.